src/Microsoft.AspNet.Razor/TagHelpers/TagHelperDescriptorComparer.cs

```diff
@@ -29,15 +30,16 @@ public class TagHelperDescriptorComparer : IEqualityComparer<TagHelperDescriptor
         /// <c>false</c> otherwise.</returns>
         /// <remarks>
         /// Determines equality based on <see cref="TagHelperDescriptor.TypeName"/>,
-        /// <see cref="TagHelperDescriptor.AssemblyName"/>, <see cref="TagHelperDescriptor.TagName"/> and
-        /// <see cref="TagHelperDescriptor.Prefix"/>.
+        /// <see cref="TagHelperDescriptor.AssemblyName"/>, <see cref="TagHelperDescriptor.TagName"/>,
+        /// <see cref="TagHelperDescriptor.Prefix"/> and <see cref="TagHelperDescriptor.RequiredAttributes"/>.
         /// </remarks>
         public bool Equals(TagHelperDescriptor descriptorX, TagHelperDescriptor descriptorY)
         {
             return string.Equals(descriptorX.TypeName, descriptorY.TypeName, StringComparison.Ordinal) &&
                    string.Equals(descriptorX.TagName, descriptorY.TagName, StringComparison.OrdinalIgnoreCase) &&
                    string.Equals(descriptorX.Prefix, descriptorY.Prefix, StringComparison.OrdinalIgnoreCase) &&
-                   string.Equals(descriptorX.AssemblyName, descriptorY.AssemblyName, StringComparison.Ordinal);
+                   string.Equals(descriptorX.AssemblyName, descriptorY.AssemblyName, StringComparison.Ordinal) &&
+                   Enumerable.SequenceEqual(descriptorX.RequiredAttributes, descriptorY.RequiredAttributes);
```

Comment:
`SequenceEqual()` requires the orders to match.  that seems overzealous here.  instead confirm counts match then compare ordered names.  or order the attribute names when constructing the `TagHelperDescriptor` instance.